pkg/skaffold/schema/v1alpha1/config.go

```diff
@@ -0,0 +1,146 @@
+type BuildType struct {
+	LocalBuild       *LocalBuild       `yaml:"local,omitempty"`
+	GoogleCloudBuild *GoogleCloudBuild `yaml:"googleCloudBuild,omitempty"`
+}
+
+// LocalBuild contains the fields needed to do a build on the local docker daemon
+// and optionally push to a repository.
+type LocalBuild struct {
+	SkipPush *bool `yaml:"skipPush,omitempty"`
+}
+
+type GoogleCloudBuild struct {
+	ProjectID string `yaml:"projectId"`
+}
+
+// DeployConfig contains all the configuration needed by the deploy steps
+type DeployConfig struct {
+	Name       string `yaml:"name,omitempty"`
+	DeployType `yaml:",inline"`
+}
+
+// DeployType contains the specific implementation and parameters needed
+// for the deploy step. Only one field should be populated.
+type DeployType struct {
+	HelmDeploy    *HelmDeploy    `yaml:"helm,omitempty"`
+	KubectlDeploy *KubectlDeploy `yaml:"kubectl,omitempty"`
+}
+
+// KubectlDeploy contains the configuration needed for deploying with `kubectl apply`
+type KubectlDeploy struct {
+	Manifests []Manifest `yaml:"manifests"`
+}
+
+type Manifest struct {
+	Paths      []string          `yaml:"paths"`
+	Parameters map[string]string `yaml:"parameters,omitempty"`
+}
+
+type HelmDeploy struct {
+	Releases []HelmRelease `yaml:"releases"`
+}
+
+type HelmRelease struct {
+	Name           string            `yaml:"name"`
+	ChartPath      string            `yaml:"chartPath"`
+	ValuesFilePath string            `yaml:"valuesFilePath"`
+	Values         map[string]string `yaml:"values"`
+	Namespace      string            `yaml:"namespace"`
+	Version        string            `yaml:"version"`
+}
+
+// Artifact represents items that need should be built, along with the context in which
+// they should be built.
+type Artifact struct {
+	ImageName      string             `yaml:"imageName"`
+	DockerfilePath string             `yaml:"dockerfilePath,omitempty"`
+	Workspace      string             `yaml:"workspace"`
+	BuildArgs      map[string]*string `yaml:"buildArgs,omitempty"`
+}
+
+// DefaultDevSkaffoldConfig is a partial set of defaults for the SkaffoldConfig
+// when dev mode is specified.
+// Each API is responsible for setting its own defaults that are not top level.
+var DefaultDevSkaffoldConfig = &SkaffoldConfig{
+	Build: BuildConfig{
+		TagPolicy: constants.DefaultDevTagStrategy,
+	},
+}
+
+// DefaultRunSkaffoldConfig is a partial set of defaults for the SkaffoldConfig
+// when run mode is specified.
+// Each API is responsible for setting its own defaults that are not top level.
+var DefaultRunSkaffoldConfig = &SkaffoldConfig{
+	Build: BuildConfig{
+		TagPolicy: constants.DefaultRunTagStrategy,
+	},
+}
+
+// Parse reads from an io.Reader and unmarshals the result into a SkaffoldConfig.
+// The default config argument provides default values for the config,
+// which can be overridden if present in the config file.
+func (config *SkaffoldConfig) Parse(contents []byte, useDefault bool, mode bool) error {
```

Comment:
yeah I went back and forth on this one, but ended up moving it there because each implementation of `Parse` uses different default values which are bound to the versioned config. I could hear arguments to move it out but to do that we'd need to have some other logic somewhere else that tells us which default to use, and I'm not sure what that would look like